src/transformers/trainer.py

```diff
@@ -1362,20 +1362,24 @@ def train(
                     self.state.best_model_checkpoint, load_optimizer_states=False, load_lr_scheduler_states=False
                 )
 
+        # add remaining tr_loss
+        self._total_loss_scalar += tr_loss.item()
+        train_loss = self._total_loss_scalar / self.state.global_step
+
         metrics = speed_metrics("train", start_time, num_samples=num_train_samples, num_steps=self.state.max_steps)
         self.store_flos()
         metrics["total_flos"] = self.state.total_flos
-        self.log(metrics)
-
-        self.control = self.callback_handler.on_train_end(args, self.state, self.control)
-        # add remaining tr_loss
-        self._total_loss_scalar += tr_loss.item()
+        metrics["train_loss"] = train_loss
 
         self.is_in_train = False
 
         self._memory_tracker.stop_and_update_metrics(metrics)
 
-        return TrainOutput(self.state.global_step, self._total_loss_scalar / self.state.global_step, metrics)
+        self.log(metrics)
+
+        self.control = self.callback_handler.on_train_end(args, self.state, self.control)
```

Comment:
Was this callback somehow coupled with :

```
        # add remaining tr_loss
        self._total_loss_scalar += tr_loss.item()
        train_loss = self._total_loss_scalar / self.state.global_step
```

i.e. is the code above (`self._total_loss_scalar`) dependent on something happening in `self.control = self.callback_handler.on_train_end`?

If it is then I need to revert this last commit and instead fix `tests/test_trainer_callback.py::TrainerCallbackTest::test_event_flow` to swap the order of:
```
        expected_events += ["on_log", "on_train_end"]
```
in `get_expected_events`.

Reply:
There is a dependency: in all reporting callbacks: `on_log` will log the last logs, then `on_train_end` will close the channel of communication between the Trainer and the integration platform (such as wandb and the like).
So the right order is on_log then on_train_end (I'm not sure if this was respected, but it should be!)

Reply:
I restored that. All looks good then. merging.